`LocalChain` with hardwired genesis block

evanlinjin opened the pull request:
closes #1079
closes #1107

### Description

Many methods of `TxGraph` require a `chain_tip: BlockId` input to use against a `ChainOracle` implementation. This is used to ask the `ChainOracle` implementation whether a certain block exists in the chain identified by the `chain_tip`. This guarantees that the `TxGraph` methods will return a consistent history of transactions.

However, the `ChainOracle` trait's `get_chain_tip` method returns an option of `BlockId`. It becomes unclear what to do when `get_chain_tip` returns `None`.

This PR changes the `ChainOracle::get_chain_tip` method to always return a `BlockId` (no `Option`). `LocalChain` now hardwires the genesis block in order to implement `ChainOracle`.

`bdk::Wallet` and `bdk_file_store::Store` are changed to have separate constructor methods for initializing a fresh instance and recovering a previous instance from persistence.

### Notes to the reviewers

### Changelog notice

- Changed `ChainOracle::get_chain_tip` method to return a `BlockId` instead of an `Option` of a `BlockId`.
- Refactored `LocalChain` so that the genesis `BlockId` is hardwired. This way, the `ChainOracle::get_chain_tip` implementation can always return a tip.
- Add `is_empty` method to `PersistBackend`. This returns true when there is no data in the persistence.
- Changed `Wallet::new` to initialize a fresh wallet only.
- Added `Wallet::load` to restore an instance of a wallet.
- Replaced `Store::new` with separate methods to create/open the database file.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

evanlinjin committed:
feat!: `LocalChain` with hardwired genesis checkpoint

This ensures that `LocalChain` will always have a tip. The `ChainOracle`
trait's `get_chain_tip` method no longer needs to return an option.

LLFourn commented on crates/bdk/README.md:
```diff
@@ -72,7 +72,7 @@ fn main() {
 
     let descriptor = "wpkh(tprv8ZgxMBicQKsPdy6LMhUtFHAgpocR8GC6QmwMSFpZs7h6Eziw3SpThFfczTDh5rW2krkqffa11UpX3XkeTTB2FvzZKWXqPY54Y6Rq4AQ5R8L/84'/0'/0'/0/*)";
-    let mut wallet = Wallet::new(descriptor, None, db, Network::Testnet).expect("should create");
+    let mut wallet = Wallet::new(descriptor, None, db, Network::Testnet, None).expect("should create");
```
I'd say either make new functions to support this or create a `Params` type or a `WalletBuilder` or something like that.  It's hard to figure out what these `None`s mean just by looking at the signature.

evanlinjin replied:
Let's do this in a separate PR, once #1101 is completed also.

LLFourn replied:
Do what? Please don't add another param to this function in this PR.

evanlinjin replied:
Sorry I misread your initial comment. I created a separate method for now.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -310,7 +358,9 @@ impl<D> Wallet<D> {
-        chain.apply_changeset(&changeset.chain);
+        chain
+            .apply_changeset(&changeset.chain)
+            .map_err(|_| NewError::InvalidPersistenceGenesis)?;
```
I think we should consider a explicit API  which distinguishes between initializing a wallet for the first time and loading it -- so `::new` and `::load`. It should be simpler for developers and could avoid this error case. Furthermore consider the implications of also doing this: https://github.com/bitcoindevkit/bdk/issues/1101#issuecomment-1701929749 and we also persist the `Network` in the first changeset (waving hands a bit here). Then load could load everything from in one go without the possibility of contradiction between runtime data and persisted data.

So `new` would take what it takes now but would persist the genesis block, network and keychains immediately. `load` would get everything from persistence and applications would just inspect it at runtime to find out what kind of wallet the user had. If we really want this kind of function we could have `new_or_load` as well which just tries to load, checks everything is consistent with the paramters (throwing errors like above if not) or if it can't be loaded, making a new one.

evanlinjin replied:
I think `Wallet::new` probably needs a check as to whether the underlying `PersistenceBackend` is empty. Otherwise, there may be accidental overwrites.

LLFourn replied:
I don't think the `PersistenceBackend` has an overwrite API though?

evanlinjin replied:
When `bdk_file_store::Store` is initialized, the logical cursor is positioned after the "magic bytes".

`Wallet::new` does not load any data from the `PersistenceBackend`.

So the next write will start at the position after "magic bytes" (if I'm not mistaken).

`Wallet::new` always writes the initial changeset.

LLFourn replied:
So this sounds like an implementation specific problem. I don't think that `Wallet::new` needs to check anything here. We can just have a safe method in `bdk_file_store` that errors out when you create a new db for a file that already exists. Remember the data we are storing is fine to delete. It can always be reconstructed.

evanlinjin replied:
I now have separate methods on `bdk_file_store::Store` for initializing and loading.

LLFourn replied:
So I didn't get feedback on the idea of doing the changes I suggested in #1101 as part of this. Is this PR intended to be a stepping stone towards doing #1101 or do you not think #1101 is viable. I doesn't seem valuable to do this change and then change the API again when we do #1101.

evanlinjin replied:
~How about we first move the descriptors to a field in the wallet's ChangeSet. Then in a separate PR, we change the keychain txout index to have changesets with descriptors? This way the API of wallet stays consistent.~

^ The above is a bad idea.

I think it's best to wait for @danielabrozzoni's PR for #1101 and have that merged first and base this work on that.

notmandatory commented on crates/chain/src/genesis.rs:
```diff
@@ -0,0 +1,54 @@
+//! Bitcoin genesis hashes.
```
Why can't we use `rust-bitcoin` for this?

https://docs.rs/bitcoin/latest/bitcoin/blockdata/constants/fn.genesis_block.html#

evanlinjin replied:
I didn't know this existed. 😮

evanlinjin replied:
Removed.

evanlinjin committed:
feat(bdk)!: have separate methods for creating and loading `Wallet`

`Wallet::new` now creates a new wallet. `Wallet::load` loads an existing
wallet. The network type is now recoverable from persistence. Error
types have been simplified.

evanlinjin committed:
feat(chain, file_store): add `is_empty` method to `PersistBackend` trait

evanlinjin committed:
chore(file_store): rm empty test file

evanlinjin committed:
feat(wallet)!: add `new_or_load` methods

These methods try to load wallet from persistence and initializes the
wallet instead if non-existant.

An internal helper method `create_signers` is added to reuse code.
Documentation is also improved.

evanlinjin commented:
I'm really close. Remaining tasks:

- [x] Make `bdk_file_store`'s constructor methods "safe" so we don't overwrite existing valid data. As mentioned here: https://github.com/bitcoindevkit/bdk/pull/1178#discussion_r1374036437
- [x] Documentation and tests for `Wallet`'s constructors.

evanlinjin committed:
feat(file_store)!: have separate methods for creating and opening Store

evanlinjin committed:
chore(examples): use `Wallet::new_or_load` method where appropriate

evanlinjin committed:
test(file_store): add construction method tests

evanlinjin committed:
test(bdk): add tests for wallet constructor methods

LLFourn commented on crates/chain/src/persist.rs:
```diff
@@ -80,6 +80,15 @@ pub trait PersistBackend<C> {
     fn load_from_persistence(&mut self) -> Result<C, Self::LoadError>;
+
+    /// Returns whether the persistence backend contains no data.
+    fn is_empty(&mut self) -> Result<bool, Self::LoadError>
```
why does this need to be on the trait. I don't see why we need it.

evanlinjin replied:
This is for the `Wallet::new_or_load` method. So that we know whether we need to initialize (where we commit the first changeset) or load from existing. Technically, we can check whether required fields exists (i.e. network type and descriptors)... Would you prefer this way instead?

**Edit:** After thinking about it, I think a better option would be for `PersistBackend::load_from_persistence` method to return `Result<Option<C>, Self::LoadError>`. It is helpful to know the difference between having no data in persistence v.s. only having persisted a `Default` changeset.

evanlinjin committed:
feat!: change `load_from_persistence` to return an option

`PersistBackend::is_empty` is removed. Instead, `load_from_persistence`
returns an option of the changeset. `None` means persistence is empty.
This is a better API than a separate method. We can now differentiate
between a persisted single changeset and nothing persisted.

`Store::aggregate_changeset` is refactored to return a `Result` instead
of a tuple. A new error type (`AggregateChangesetsError`) is introduced
to include the partially-aggregated changeset in the error. This is a
more idiomatic API.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -126,12 +126,18 @@ pub struct ChangeSet {
     fn append(&mut self, other: Self) {
         Append::append(&mut self.chain, other.chain);
         Append::append(&mut self.indexed_tx_graph, other.indexed_tx_graph);
+        if other.network.is_some() {
```
Should probably `debug_assert` at least that these are not conflicting.

A cleaner less brittle approach would be to have some fixed `Header` data that you can set in persistence. Make an issue if you think it's worth discussing.

LLFourn reviewed (approved):
ACK 632b8067cd173700c04aa1b848da49e13c0dea13

evanlinjin committed:
feat(bdk): changeset's `Append` impl checks that network is consistent

vladimirfomene commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -43,29 +43,12 @@ pub struct Emitter<'c, C> {
-    ///
-    /// `start_height` is the block height to start emitting blocks from.
-    pub fn from_height(client: &'c C, start_height: u32) -> Self {
+    /// TODO
```
I wonder what we can do so that we don't forget to write this doc.

notmandatory replied:
Looks like docs here should be something like:
```suggestion
    /// Construct a new [`Emitter`] with the given RPC `client`, `last_cp` and `start_height`.
    ///
    /// `last_cp` is the check point used to find the latest block which is still part of the best chain.
    /// `start_height` is the block height to start emitting blocks from.
```

vladimirfomene commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -221,99 +231,215 @@ impl Wallet {
 impl<D> Wallet<D> {
-    /// Create a wallet from a `descriptor` (and an optional `change_descriptor`) and load related
-    /// transaction data from `db`.
+    /// Initialize an empty [`Wallet`].
```
What does `empty` here mean? From my understanding, it means the wallet has an empty state (empty IndexedTxGraph, empty localchain). We might want to improve the doc here.

vladimirfomene commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -326,6 +452,143 @@ impl<D> Wallet<D> {
+        Self::load_from_changeset(descriptor, change_descriptor, db, changeset)
+    }
+
+    fn load_from_changeset<E: IntoWalletDescriptor>(
```
nits: I will change the name of this method to `init_from_changeset`. Feel free to ignore if it doesn't resonate.

vladimirfomene commented on crates/file_store/src/store.rs:
```diff
@@ -33,30 +33,64 @@ where
+    C: Append + serde::Serialize + serde::de::DeserializeOwned,
 {
-    /// Creates a new store from a [`File`].
+    /// Create a new [`Store`] file in write-only mode; error if the file exists.
```
Is it really in write only mode given that the file is also given read option on line 65

vladimirfomene reviewed (commented):
Great work Evan! I just have some nits.

notmandatory reviewed (approved):
ACK e342a8c3dd5ce7775756df8ae7bbbe2eb3f61863

Looks good, nice to have that chain tip Option removed!

notmandatory commented:
This should also close #1107 .

notmandatory commented:
@evanlinjin looks like this one has a conflict so needs a rebase before I can merge it.

evanlinjin commented:
@notmandatory don't merge yet, I want to address comments and make clippy happier.

evanlinjin committed:
chore: make clippy happy

evanlinjin committed:
docs(bitcoind_rpc): update docs for `Emitter::new`
